File: apps/docker-server/cmd/renewlet/notification_message_storage_test.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
	"time"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

func TestNotificationHistoryBrowserFixture(test *testing.T) {
	userID := os.Getenv("RENEWLET_E2E_NOTIFICATION_USER")
	var app *pocketbase.PocketBase
	if userID == "" {
		app = newSchemaTestApp(test)
		if err := ensureSchema(app); err != nil {
			test.Fatal(err)
		}
		user, _ := createRouteTestUser(test, app, "notification-browser-fixture")
		userID = user.Id
	} else {
		// 浏览器夹具只允许写 Playwright 已创建的固定隔离库；不接收数据目录，也不负责建库或迁移。
		directory, err := filepath.Abs("../../pb_data_e2e")
		if err != nil {
			test.Fatal(err)
		}
		resolved, err := filepath.EvalSymlinks(directory)
		if err != nil || resolved != directory {
			test.Fatalf("browser fixture requires a non-symlink E2E directory: %v", err)
		}
		info, err := os.Lstat(filepath.Join(directory, "data.db"))
		if err != nil || !info.Mode().IsRegular() {
			test.Fatalf("browser fixture requires the existing E2E database: %v", err)
		}
		app = pocketbase.NewWithConfig(pocketbase.Config{DefaultDataDir: directory})
		registerAuthHooks(app)
		if err := app.Bootstrap(); err != nil {
			test.Fatal(err)
		}
		test.Cleanup(func() { _ = app.ResetBootstrapState() })
		if _, err := app.FindRecordById("users", userID); err != nil {
			test.Fatal(err)
		}
	}
	registerRecordHooks(app)
	expectedMessages := make(map[string]notificationJobResultMessage, 12)
	// 整批夹具走同一真实事务和 finalize，历史 API 必须读到完整正文，而不是由浏览器模拟响应。
	if err := app.RunInTransaction(func(txApp core.App) error {
		for index := 0; index < 12; index++ {
			result := largeNotificationJobResult(1)
			instant := time.Date(2026, 5, 18, 20, index, 0, 0, time.UTC).Format(time.RFC3339)
			result.Schedule = localScheduleOccurrence{
				ScheduledLocalDate: "2026-05-19", ScheduledLocalTime: fmt.Sprintf("04:%02d", index),
				TimeZone: "Asia/Shanghai", ScheduledInstantUTC: instant,
			}
			result.TriggeredAtUTC = instant
			result.Settings.Timezone = "Asia/Shanghai"
			result.Settings.NotificationTimeLocal = "04:00"
			result.Settings.EnabledChannels = []string{"email"}
			result.Message.Title = "Renewlet 订阅提醒"
			result.Message.Timestamp = instant
			result.Message.Content = fmt.Sprintf("通知内容快照 %d\n%s", index, strings.Repeat("Long diagnostic payload ", 40))
			result.Message.Items[0].Name = fmt.Sprintf("Notification Drawer Seed %d", index)
			result.Message.Items[0].Price = "19"
			result.Message.Items[0].Currency = "USD"
			result.Message.Items[0].TargetDate = "2026-05-20"
			result.Message.Items[0].ReminderDays = 1
			result.Message.Items[0].DaysUntil = 1
			reason := "some_channels_failed"
			result.Reason = &reason
			errorText := fmt.Sprintf("email: dial tcp: lookup smtp.example.com: no such host %d", index)
			result.Channels = jobChannels{Attempted: []string{"email"}, Succeeded: []string{}, Failed: []channelFailure{{Channel: "email", Error: errorText}}}
			schedule := localScheduleDecision{localScheduleOccurrence: result.Schedule}
			job, _, err := createNotificationJob(txApp, userID, schedule, notificationStatusFailed, index+1)
			if err != nil {
				return err
			}
			if job == nil {
				return fmt.Errorf("browser notification fixture %d already exists", index)
			}
			if err := finalizeNotificationJob(txApp, job, userID, schedule, notificationStatusFailed, errorText, result); err != nil {
				return err
			}
			expectedMessages[job.Id] = result.Message
		}
		return nil
	}); err != nil {
		test.Fatal(err)
	}
	history, err := loadNotificationHistoryJobs(app, userID, "all", 20, 0)
	if err != nil || len(history) != len(expectedMessages) {
		test.Fatalf("browser fixture history is incomplete: %v", err)
	}
	for _, job := range history {
		if job.Status != notificationStatusFailed || !reflect.DeepEqual(assertNormalizedCronResult(test, job).Message, expectedMessages[job.ID]) {
			test.Fatal("browser fixture changed message or final state")
		}
	}
}

func TestNotificationLargeHistoryPreservesCompleteMessage(t *testing.T) {
	for _, size := range []int{10, 100, 1000, 5000} {
		t.Run(fmt.Sprint(size), func(t *testing.T) {
			app := newSchemaTestApp(t)
			if err := ensureSchema(app); err != nil {
				t.Fatal(err)
			}
			user, token := createRouteTestUser(t, app, "large-history")
			result := largeNotificationJobResult(size)
			schedule := localScheduleDecision{localScheduleOccurrence: result.Schedule}
			record, _, err := createNotificationJob(app, user.Id, schedule, notificationStatusSending, 1)
			if err != nil {
				t.Fatal(err)
			}
			payload, err := json.Marshal(result)
			if err != nil {
				t.Fatal(err)
			}
			t.Logf("items=%d complete result bytes=%d", size, len(payload))
			// 最终状态保存不能受消息条数影响；历史正文必须保留发送时快照，不能截断或从当前订阅重建。
			if err := finalizeNotificationJob(app, record, user.Id, schedule, notificationStatusSent, "", result); err != nil {
				t.Fatal(err)
			}
			history := requestNotificationHistory(t, app, token)
			if len(history.Jobs) != 1 || history.Jobs[0].Status != notificationStatusSent {
				t.Fatalf("unexpected completed history: %+v", history)
			}
			actual := assertNormalizedCronResult(t, history.Jobs[0])
			if !reflect.DeepEqual(actual, result) {
				t.Fatal("history changed the complete sent result")
			}
			operations, err := measureSubscriptionDBOperations(app, func() error {
				_, err := loadNotificationHistoryJobs(app, user.Id, "all", 20, 0)
				return err
			})
			if err != nil || operations.ReadQueries != 1 || operations.WriteStatements != 0 {
				t.Fatalf("history must read one consistent snapshot without writes: %+v %v", operations, err)
			}
		})
	}
}

func TestNotificationMessageReplacementRollsBack(t *testing.T) {
	for _, failure := range []string{"message", "status"} {
		t.Run(failure, func(t *testing.T) {
			app := newSchemaTestApp(t)
			if err := ensureSchema(app); err != nil {
				t.Fatal(err)
			}
			user, _ := createRouteTestUser(t, app, "authenticated")
			result := largeNotificationJobResult(1000)
			schedule := localScheduleDecision{localScheduleOccurrence: result.Schedule}
			job, _, err := createNotificationJob(app, user.Id, schedule, notificationStatusSending, 1)
			if err != nil {
				t.Fatal(err)
			}
			if err := finalizeNotificationJob(app, job, user.Id, schedule, notificationStatusFailed, "old failure", result); err != nil {
				t.Fatal(err)
			}
			before, err := loadNotificationHistoryJobs(app, user.Id, "all", 1, 0)
			if err != nil {
				t.Fatal(err)
			}
			statement := "CREATE TRIGGER fail_snapshot BEFORE INSERT ON notification_job_messages BEGIN SELECT RAISE(ABORT, 'injected snapshot failure'); END"
			if failure == "status" {
				statement = "CREATE TRIGGER fail_snapshot BEFORE UPDATE ON notification_jobs BEGIN SELECT RAISE(ABORT, 'injected snapshot failure'); END"
			}
			if _, err := app.DB().NewQuery(statement).Execute(); err != nil {
				t.Fatal(err)
			}
			// 同时保护分段写入中断和最终态写入失败；不能只回滚其中一张表。
			if err := finalizeNotificationJob(app, job, user.Id, schedule, notificationStatusSent, "", largeNotificationJobResult(5000)); err == nil || !strings.Contains(err.Error(), "injected snapshot failure") {
				t.Fatalf("expected original failure: %v", err)
			}
			after, err := loadNotificationHistoryJobs(app, user.Id, "all", 1, 0)
			if err != nil || !reflect.DeepEqual(before, after) {
				t.Fatalf("failed replacement changed history: %v", err)
			}
		})
	}
}

func TestNotificationMessageOwnershipAndDeletion(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	user, token := createRouteTestUser(t, app, "authenticated")
	other, otherToken := createRouteTestUser(t, app, "other")
	result := largeNotificationJobResult(1000)
	schedule := localScheduleDecision{localScheduleOccurrence: result.Schedule}
	job, _, err := createNotificationJob(app, user.Id, schedule, notificationStatusSending, 1)
	if err != nil {
		t.Fatal(err)
	}
	if err := finalizeNotificationJob(app, job, other.Id, schedule, notificationStatusSent, "", result); err == nil {
		t.Fatal("accepted other user's job")
	}
	if err := finalizeNotificationJob(app, job, user.Id, schedule, notificationStatusSent, "", result); err != nil {
		t.Fatal(err)
	}
	if len(requestNotificationHistory(t, app, otherToken).Jobs) != 0 || len(requestNotificationHistory(t, app, token).Jobs) != 1 {
		t.Fatal("history owner isolation failed")
	}
	if _, err := app.DB().NewQuery("DELETE FROM notification_job_messages WHERE job_id = {:id} AND chunk_index = 1").Bind(dbx.Params{"id": job.Id}).Execute(); err != nil {
		t.Fatal(err)
	}
	if _, err := loadNotificationHistoryJobs(app, user.Id, "all", 1, 0); err == nil {
		t.Fatal("incomplete message was hidden as an empty result")
	}
	if err := app.Delete(user); err != nil {
		t.Fatal(err)
	}
	var remaining struct {
		Count int `db:"count"`
	}
	if err := app.DB().NewQuery("SELECT COUNT(*) AS count FROM notification_job_messages").One(&remaining); err != nil || remaining.Count != 0 {
		t.Fatalf("private message survived user deletion: %+v %v", remaining, err)
	}
}

func TestNotificationCronLargeBatchRetriesOnlyFailedChannel(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	registerRecordHooks(app)
	user, token := createRouteTestUser(t, app, "authenticated")
	settings := defaultAppSettings()
	settings.Timezone = "UTC"
	settings.NotificationTimeLocal = "08:00"
	settings.NotificationReminderDays = 3
	settings.EnabledChannels = []string{"webhook", "telegram"}
	settings.WebhookURL = "https://example.com/notification"
	settings.TelegramBotToken = "123:fixture"
	settings.TelegramChatID = "123"
	createNotificationCronRouteTestSettings(t, app, user, settings)
	if err := app.RunInTransaction(func(txApp core.App) error {
		for index := 0; index < 1000; index++ {
			createRouteTestSubscription(t, txApp, user.Id, map[string]interface{}{"name": fmt.Sprintf("Large batch %d", index), "autoRenew": false, "nextBillingDate": "2026-09-11", "reminderDays": 3})
		}
		return nil
	}); err != nil {
		t.Fatal(err)
	}
	webhookCalls, telegramCalls := 0, 0
	failTelegram := true
	restore := withNotificationHTTPClient(t, serverChanRoundTripFunc(func(request *http.Request) (*http.Response, error) {
		if request.URL.Host == "example.com" {
			webhookCalls++
			return serverChanTestResponse(http.StatusOK, `{}`), nil
		}
		if request.URL.Host != "api.telegram.org" {
			t.Fatalf("unexpected external request %s", request.URL.Host)
		}
		telegramCalls++
		if failTelegram {
			return serverChanTestResponse(http.StatusBadRequest, `{"ok":false,"description":"fixture failure"}`), nil
		}
		return serverChanTestResponse(http.StatusOK, `{"ok":true}`), nil
	}))
	defer restore()
	options := notificationCronOptions{Now: time.Date(2026, 9, 8, 8, 0, 0, 0, time.UTC), WindowMinutes: 2, MaxRetries: 3, StaleSendingMinutes: 15}
	refreshNotificationSchedulerForTest(t, app, user.Id, options.Now)
	first, err := runNotificationCron(app, options)
	if err != nil || first.Failed != 1 {
		t.Fatalf("first run: %+v %v", first, err)
	}
	failTelegram = false
	second, err := runNotificationCron(app, options)
	if err != nil || second.Sent != 1 {
		t.Fatalf("retry run: %+v %v", second, err)
	}
	completedTelegramCalls := telegramCalls
	third, err := runNotificationCron(app, options)
	if err != nil || third.Sent != 0 || third.Failed != 0 || webhookCalls != 1 || telegramCalls != completedTelegramCalls {
		t.Fatalf("successful channel resent: webhook=%d telegram=%d result=%+v err=%v", webhookCalls, telegramCalls, third, err)
	}
	history := requestNotificationHistory(t, app, token)
	if len(history.Jobs) != 1 || len(assertNormalizedCronResult(t, history.Jobs[0]).Message.Items) != 1000 {
		t.Fatal("large cron lost history items")
	}
}

func largeNotificationJobResult(size int) notificationJobResult {
	items := make([]notificationContentItem, size)
	for index := range items {
		items[index] = notificationContentItem{
			Type: "renewal", SubscriptionID: fmt.Sprintf("sub%012d", index), Name: fmt.Sprintf("订阅 🌏 %d", index),
			Price: "12.50", Currency: "CNY", Status: "active", TargetDate: "2026-09-11", ReminderDays: 3, DaysUntil: 3,
		}
	}
	settings := defaultAppSettings()
	settings.Timezone = "UTC"
	settings.NotificationTimeLocal = "08:00"
	settings.EnabledChannels = []string{"webhook"}
	now := time.Date(2026, 9, 8, 8, 0, 0, 0, time.UTC)
	schedule := localScheduleOccurrence{ScheduledLocalDate: "2026-09-08", ScheduledLocalTime: "08:00", TimeZone: "UTC", ScheduledInstantUTC: "2026-09-08T08:00:00Z"}
	return createJobResult("", schedule, settings, appLocale("zh-CN"), buildNotificationContent(now, settings, items, appLocale("zh-CN")), notificationCronOptions{Now: now, WindowMinutes: 2}, jobChannels{Attempted: []string{"webhook"}, Succeeded: []string{"webhook"}})
}

File: apps/worker/src/notification-message-storage.test.ts
import { readFileSync, readdirSync } from "node:fs";
import { DatabaseSync, type SQLInputValue } from "node:sqlite";
import { afterEach, describe, expect, it } from "vitest";
import { createDefaultAppSettings } from "@renewlet/shared/settings-defaults";
import { cronJobResultResponseSchema } from "@renewlet/shared/schemas/notifications";
import { createCronJobResult, createNotificationJob, finalizeNotificationJob, getNotificationJob, readJobChannels } from "./notification-jobs";
import { readNotificationHistoryRows, splitNotificationJobMessage } from "./notification-message-storage";
import type { Env } from "./types";
import upgradeFixtures from "../../../packages/shared/src/contract-fixtures/notification-message-upgrade-fixtures.json";

const migration = readFileSync(new URL("../migrations/0041_exclusive_notification_message_snapshots.sql", import.meta.url), "utf8");
const databases: DatabaseSync[] = [];
afterEach(() => { for (const database of databases.splice(0)) database.close(); });

function databaseFixture(migrate = true) {
  const database = new DatabaseSync(":memory:");
  databases.push(database);
  for (const name of readdirSync(new URL("../migrations/", import.meta.url)).filter((name) => name.endsWith(".sql") && name < "0041_exclusive_notification_message_snapshots.sql").sort()) {
    database.exec(readFileSync(new URL(`../migrations/${name}`, import.meta.url), "utf8"));
  }
  database.exec(`INSERT INTO users (id, email, name, role, password_hash, created_at, updated_at)
    VALUES ('owner', 'owner@example.test', 'Owner', 'user', '', '', ''), ('other', 'other@example.test', 'Other', 'user', '', '', '')`);
  if (migrate) database.exec(migration);
  const queries: string[] = [];
  // 本地真实 SQLite 执行相同 SQL 和事务；不是 D1 元数据或云端性能模拟结果。
  function prepare(sql: string, params: SQLInputValue[] = []): D1PreparedStatement {
    return {
      bind: (...values: SQLInputValue[]) => prepare(sql, values),
      all: async () => { queries.push(sql); return { success: true, results: database.prepare(sql).all(...params), meta: {} }; },
      first: async () => { queries.push(sql); return database.prepare(sql).get(...params) ?? null; },
      run: async () => { queries.push(sql); const result = database.prepare(sql).run(...params); return { success: true, results: [], meta: { changes: Number(result.changes) } }; },
    } as unknown as D1PreparedStatement;
  }
  const env = { DB: {
    prepare,
    batch: async (statements: D1PreparedStatement[]) => {
      database.exec("BEGIN");
      try {
        const results: D1Result[] = [];
        for (const statement of statements) results.push(await statement.run());
        database.exec("COMMIT");
        return results;
      } catch (error) { database.exec("ROLLBACK"); throw error; }
    },
  } as unknown as D1Database, ASSETS: {} as Fetcher, ASSETS_BUCKET: {} as R2Bucket } satisfies Env;
  return { database, env, queries };
}

function largeResult(size: number) {
  const settings = createDefaultAppSettings();
  settings.timezone = "UTC";
  settings.enabledChannels = ["webhook"];
  return cronJobResultResponseSchema.parse(createCronJobResult({
    reason: null, force: false, windowMinutes: 2, triggeredAtUtc: "2026-09-08T08:00:00Z",
    schedule: { scheduledLocalDate: "2026-09-08", scheduledLocalTime: "08:00", timeZone: "UTC", scheduledInstantUtc: "2026-09-08T08:00:00Z" },
    settings, locale: "zh-CN",
    message: {
      title: "Renewlet", content: "中文🌏\n".repeat(size), timestamp: "2026-09-08 08:00 UTC", hasPayload: true,
      items: Array.from({ length: size }, (_, index) => ({ type: "renewal", subscriptionId: `sub${index}`, name: `订阅🌏${index}`, price: "12.50", currency: "CNY", status: "active", targetDate: "2026-09-11", reminderDays: 3, daysUntil: 3 })),
    },
    channels: { attempted: ["webhook"], succeeded: ["webhook"], failed: [] },
  }));
}

describe("notification message snapshot storage", () => {
  it.each(upgradeFixtures.fixtures)("converts the actual $version result without changing message text", async (fixture) => {
    const { env, database } = databaseFixture(false);
    const { row } = await createNotificationJob(env, "owner", fixture.result.schedule, "failed", 2);
    database.prepare("UPDATE notification_jobs SET result_json = ? WHERE id = ?").run(JSON.stringify(fixture.result), row?.id ?? "");
    database.exec(`BEGIN; ${migration} COMMIT;`);
    const expected = structuredClone(fixture.result);
    for (const item of expected.message.items) { if (typeof item.price === "number") item.price = "12.5"; }
    const [history] = await readNotificationHistoryRows(env, "owner", "all", 20);
    expect(JSON.parse(history?.result_json ?? "{}")).toEqual(expected);
    expect(cronJobResultResponseSchema.safeParse(JSON.parse(history?.result_json ?? "{}"))).toMatchObject({ success: true });
    expect(readJobChannels(await getNotificationJob(env, "owner", fixture.result.schedule))).toEqual(fixture.result.channels);
  });

  it("rolls back a failed historical conversion before publishing the new storage", async () => {
    const { env, database } = databaseFixture(false);
    const original = largeResult(1000);
    const { row } = await createNotificationJob(env, "owner", original.schedule, "failed", 2);
    database.prepare("UPDATE notification_jobs SET result_json = ? WHERE id = ?").run(JSON.stringify(original), row?.id ?? "");
    database.exec("CREATE TRIGGER fail_upgrade BEFORE UPDATE ON notification_jobs BEGIN SELECT RAISE(ABORT, 'injected upgrade failure'); END");
    database.exec("BEGIN");
    expect(() => database.exec(migration)).toThrow("injected upgrade failure");
    database.exec("ROLLBACK");
    expect(JSON.parse((await getNotificationJob(env, "owner", original.schedule))?.result_json ?? "{}")).toEqual(original);
    expect(database.prepare("SELECT name FROM sqlite_master WHERE name = 'notification_job_messages'").get()).toBeUndefined();
    database.exec("DROP TRIGGER fail_upgrade");
    database.exec(`BEGIN; ${migration} COMMIT;`);
    expect(JSON.parse((await readNotificationHistoryRows(env, "owner", "all", 20))[0]?.result_json ?? "{}")).toEqual(original);
  });

  it.each([10, 100, 1000, 5000])("persists all %i items without increasing the job budget", async (size) => {
    const { env, database, queries } = databaseFixture();
    const result = largeResult(size);
    const { row } = await createNotificationJob(env, "owner", result.schedule, "sending", 1);
    await finalizeNotificationJob(env, row, "owner", result.schedule, "sent", 1, null, result);
    const stored = await getNotificationJob(env, "owner", result.schedule);
    expect(stored?.status).toBe("sent");
    expect(new TextEncoder().encode(stored?.result_json).length).toBeLessThanOrEqual(65536);
    expect(stored && JSON.parse(stored.result_json)).not.toHaveProperty("message");
    queries.length = 0;
    const history = await readNotificationHistoryRows(env, "owner", "all", 20);
    expect(queries).toHaveLength(1);
    expect(JSON.parse(history[0]?.result_json ?? "{}")).toEqual(result);
    expect(await readNotificationHistoryRows(env, "other", "all", 20)).toEqual([]);
    expect(() => database.prepare("UPDATE notification_jobs SET result_json = ? WHERE user_id = 'owner'").run(JSON.stringify(result))).toThrow("NOTIFICATION_MESSAGE_STORAGE_CONTRACT_INVALID");
    database.prepare("DELETE FROM users WHERE id = ?").run("owner");
    expect(database.prepare("SELECT COUNT(*) AS count FROM notification_job_messages").get()?.["count"]).toBe(0);
  });

  it("rolls back snapshot replacement when the final status write fails", async () => {
    const { env, database } = databaseFixture();
    const original = largeResult(1000);
    const { row } = await createNotificationJob(env, "owner", original.schedule, "sending", 1);
    await finalizeNotificationJob(env, row, "owner", original.schedule, "failed", 1, "failed", original);
    database.exec(`CREATE TRIGGER fail_final_status BEFORE UPDATE ON notification_jobs BEGIN SELECT RAISE(ABORT, 'injected final failure'); END`);
    await expect(finalizeNotificationJob(env, row, "owner", original.schedule, "sent", 2, null, largeResult(5000))).rejects.toThrow("injected final failure");
    expect((await getNotificationJob(env, "owner", original.schedule))?.status).toBe("failed");
    expect(JSON.parse((await readNotificationHistoryRows(env, "owner", "all", 20))[0]?.result_json ?? "{}")).toEqual(original);
  });

  it("converts the old stored shape atomically and keeps the public snapshot unchanged", async () => {
    const { env, database } = databaseFixture(false);
    const original = largeResult(1000);
    const { row } = await createNotificationJob(env, "owner", original.schedule, "failed", 2);
    database.prepare("UPDATE notification_jobs SET result_json = ? WHERE id = ?").run(JSON.stringify(original), row?.id ?? "");
    database.exec(`BEGIN; ${migration} COMMIT;`);
    const history = await readNotificationHistoryRows(env, "owner", "all", 20);
    expect(JSON.parse(history[0]?.result_json ?? "{}")).toEqual(original);
    expect(readJobChannels(await getNotificationJob(env, "owner", original.schedule))).toEqual(original.channels);
    expect(() => database.exec(migration)).toThrow();
    expect(await readNotificationHistoryRows(env, "owner", "all", 20)).toEqual(history);
  });

  it("does not accept a missing message part as an empty historical result", async () => {
    const { env, database } = databaseFixture();
    const original = largeResult(1000);
    const { row } = await createNotificationJob(env, "owner", original.schedule, "sending", 1);
    await finalizeNotificationJob(env, row, "owner", original.schedule, "sent", 1, null, original);
    expect(splitNotificationJobMessage(original).parts.length).toBeGreaterThan(1);
    database.exec("DELETE FROM notification_job_messages WHERE chunk_index = 1");
    await expect(readNotificationHistoryRows(env, "owner", "all", 20)).rejects.toThrow("missing part");
  });
});
